doc/parallel.md

```diff
@@ -0,0 +1,67 @@
+Parallel testing
+=========================
+Standard RapidCheck stateful testing can be used to find bugs that occur as a result of single threaded interaction with the system under test. However, another very common class of bugs are race conditions which by definition only appear when more than one thread is involved. RapidCheck can provoke race conditions in stateful tests by executing a subset of the commands in a command sequence in parallel. A parallel test is conseridered successful if there is a possible interleaving of the commands in the test that can explain the result of the execution.
+
+## The anatomy of a parallel test ##
+The execution of a parallel test case can be divided into four phases: command generation, pre-condition verification, execution and verification. 
+
+During command generation three distinct command sequences are generated: the prefix, the left branch and the right branch. The prefix will be executed first, followed by the left and right branches that are executed in parallel. The exact order in which the commands in the left and right branches will be executed is not known to RapidCheck. This makes pre-condition verification more strict than in the sequential case as pre-condition verification for each test case has to be done for all possible command interleavings. The test case is discarded if any failing pre-condition is found.
+
+After pre-condition verification comes the execution. As explained earlier RapidCheck first executes the prefix, followed by concurrently executing the left and right branches in different threads. Once the execution is completed, RapidCheck checks if there exists at least one valid (linearized) command interleaving. A valid command interleaving is a possible ordering of commands for which there are no failing assertions when the validation functions that were returned from the ```run``` function of each command in the sequence are invoked sequentially.
```

Comment:
How about moving this section to the bottom and adding a disclaimer like: "Users of RapidCheck normally don't have to know the exact details of how RapidCheck executes parallel tests. But for those that are interested this section contains an overview of RapidCheck's' parallel testing implementation."